### PR TITLE
fix(reports): make report drill-down filter in Advanced mode

### PR DESCRIPTION
**Pull request by carlosvirreira**
## Need

Customer (IT supervisor at a UK org, ~5k assets / 207 locations) reported that clicking a slice on the **Asset Distribution** donut took him to the Assets index, but the list was **not filtered** to the slice he clicked. Confirmed via live reproduction on a workspace in Advanced mode: clicking "Storage Room Second Floor" emitted `/assets?location=<uuid>`, the server stripped the param via a redirect, and the user landed on the unfiltered index.

The drill-down works in Simple mode. The bug is exclusively in Advanced mode.

## Hypothesis (confirmed)

`getAdvancedFiltersFromRequest` in `apps/webapp/app/utils/cookies.server.ts` validates each column filter against `advancedFilterFormatSchema`, which requires the operator form (`is:<value>`). Bare values (the form everything historically emits — report drill-down, external deep links, hard-coded `?status=ONGOING` links) failed validation and were **silently dropped**. The loader then set `redirectNeeded = true` and redirected to a URL built from the cookie state, clobbering the user's intent. The same fix path also unblocked Custody Snapshot's `locationId` parameter, which the route was plumbing in but the helper was destructuring and ignoring.

## Changes

- **`cookies.server.ts`** — Validator now **normalizes** bare column-filter values to `is:<value>` instead of dropping them. Operator-prefixed values pass through unchanged. **Values containing `:` that fail operator validation are treated as malformed operator attempts and dropped** (preserving the pre-PR drop-on-fail semantics), so something like `?status=foo:AVAILABLE` cannot get coerced into `is:foo:AVAILABLE` and crash the downstream enum cast. The CASE 1 + CASE 2 loops share a single private `normalizeAdvancedFilterParams` helper. The redirect still fires when normalization changed the params, preserving cookie persistence and URL shareability.

- **`helpers.server.ts`** — `custodySnapshotReport` now destructures and applies `locationId`. The function's interface had been declaring the parameter and the route had been plumbing it in (from `?location=<uuid>`), but the function body silently dropped it. The `without-location` sentinel maps to `locationId IS NULL`, matching Simple-mode convention.

- **`distribution-donut.tsx`** — No defensive changes. The asset query layer (`query.server.ts:492-993`) already translates the `without-location` / `uncategorized` / `without-custody` sentinels to `IS NULL` SQL, so the donut handler's bare-value URLs flow through correctly once the normalizer prepends `is:`.

## Test plan

- [x] 7 new vitest unit tests (`cookies.server.test.ts`) covering: operator pass-through, bare promotion, multi-dim normalization (location/category/status), non-column param passthrough, empty-value handling, cookie-empty case.
- [x] All 2004 vitest tests still pass.
- [x] `pnpm webapp:typecheck` clean.
- [x] Manual repro in Advanced mode: clicking "Storage Room Second Floor" slice → URL becomes `/assets?location=is%3A<uuid>`, list filters to 1 asset, "Storage Room Second Floor" chip appears.
- [x] Manual regression in Simple mode: same click → URL `/assets?location=<uuid>` (unchanged), 1 asset, chip visible.
- [x] Sentinel drill-down in Advanced mode: clicking "No Location" → URL `/assets?location=is:without-location`, list filters to the 1 asset whose `locationId` is null. (Backend has `IS NULL` translation since before this PR; the normalizer makes the donut's bare-value URL reach it.)
- [x] Custody Snapshot manual checks:
  - `/reports/custody-snapshot` → 1 row (the asset in custody)
  - `/reports/custody-snapshot?location=<storage-room-uuid>` → 0 rows (asset is not at that location, correctly filtered)
  - `/reports/custody-snapshot?location=without-location` → 1 row (sentinel → IS NULL branch)

## Cross-report audit (separate sub-agents)

I ran a parallel audit of all 10 reports to find adjacent gaps of the same nature. Findings:

| Report | Status |
|---|---|
| Asset Distribution (this PR) | ✅ Fixed |
| Custody Snapshot (this PR) | ✅ Fixed |
| Asset Inventory, Idle Assets, Asset Activity, Asset Utilization, Booking Compliance, Top Booked Assets, Monthly Trends, Overdue Items | Navigate to detail pages (no filter params emitted) — unaffected |
| `at-risk-bookings.tsx` | Emits bare `?status=ONGOING` to `/bookings`. **Works today** because Bookings has no Advanced mode. Tracked in #2543 as a contract-inheritance constraint for any future Bookings Advanced-mode PR. |

## Follow-ups filed

- **#2543** — Bookings Advanced-mode contract inheritance. If/when Bookings grows an Advanced mode, the bare-value normalization contract must inherit; `at-risk-bookings.tsx` will otherwise break the same way.
- **#2542** — Reports capability-gap design pass. Asset Utilization bar-chart drill-down, Custody Snapshot custodian column → custodian detail, Monthly Trends month → bookings drill-down, etc. Not bugs, capability nits — needs a design decision before implementation.

## Unrelated CI noise

- **#2544** — React Doctor sticky comment misreports clean runs as "scan may have failed". The actual check is green on this PR (`100/100`); only the bot's comment is misleading. Workflow logic bug at `.github/workflows/react-doctor.yml:91`. Not blocking.

## Out of scope

- The two-mode Simple/Advanced architecture itself. The normalizer change makes the URL grammar mode-agnostic for the common case (bare equality), which is most of what's worth fixing here. A bigger UX call (kill the toggle entirely? progressive operator disclosure à la Linear/Notion?) deserves its own design conversation.

## Summary by CodeRabbit

* **Bug Fixes**
  * Custody snapshot reports now correctly filter assets by their assigned location, improving report accuracy.
  * Advanced filter parameters are now properly normalized and validated, with improved handling of various input formats and edge cases.

* **Tests**
  * Added comprehensive test coverage for advanced filter parameter normalization and validation logic to ensure consistent behavior.

[Review Change Stack](https://app.coderabbit.ai/change-stack/Shelf-nu/shelf.nu/pull/2540)

### CODE REVIEW

**Comment by carlosvirreira** [comment]
@coderabbitai review

File: apps/webapp/app/components/reports/distribution-donut.tsx [hunk 1]
```diff
@@ -157,48 +174,51 @@ export function DistributionDonut({
...
-                  {item.assetCount}
-                </span>
-                <span className="text-xs text-gray-500">
-                  ({item.percentage.toFixed(0)}%)
-                </span>
-              </div>
-            </button>
-          ))}
+          {legendItems.map((item, index) => {
+            const itemClickable = isClickable && !SENTINEL_IDS.has(item.id);
```
[the remark this answers is not in the document]
Comment (carlosvirreira) [reply apps/webapp/app/components/reports/distribution-donut.tsx]: Great catch — you're right. I traced through `query.server.ts` after seeing this and confirmed the backend already handles all three sentinels (`uncategorized`, `without-location`, `without-custody`) by translating them to `IS NULL` SQL (lines 492-526 / 564-614 / 889-993).

The `SENTINEL_IDS` gate I added was defensive code based on a wrong assumption that Advanced mode lacked null-filter semantics. Removed in commit f091b0efb. Verified end-to-end that clicking "No Location" in Advanced mode now correctly drills into `/assets?location=is:without-location` and filters to assets with `locationId IS NULL`.

Closed the misdirected follow-up #2541. Thanks.

**Comment by carlosvirreira** [comment]
Round-2 changes pushed in f091b0efb. Addressing each piece of review feedback:

**@chatgpt-codex-connector (P2) — "Keep sentinel legend items clickable"**

✅ Fixed. You were right: `query.server.ts` already handles all three sentinels (`uncategorized` / `without-location` / `without-custody`) as `IS NULL` SQL (lines 492-526 / 564-614 / 889-993). The gate I added was defensive code based on a wrong assumption. Removed entirely; verified end-to-end that the donut "No Location" slice now drills correctly into `/assets?location=is:without-location` → 1 asset filtered. Closed the misdirected follow-up #2541.

**@coderabbitai nit 1 — Centralize sentinel IDs in shelf.config.ts**

✅ Moot after the above — the sentinel set was removed entirely. Nothing to centralize.

**@coderabbitai nit 2 — Remove unnecessary `as any` cast**

✅ Fixed. Dropped at both call sites. `columnNames: string[]` and the `URLSearchParams.forEach` key arg are both plain strings, so `includes()` works without the cast.

**@coderabbitai nit 3 — Extract duplicated normalization logic to a helper**

✅ Fixed. Pulled the CASE 1 + CASE 2 loops into a private `normalizeAdvancedFilterParams(filters, columnNames)` helper. Single source of truth; both call sites now one-liner through the helper. Also hoisted `columnNames` computation above the case split so it's not recomputed.

**Verification of round-2 changes:**

- All 2004 vitest tests still pass (the 7 normalizer tests are unchanged in behavior).
- `pnpm webapp:typecheck` clean.
- Manual browser repro in Advanced mode: clicking "Storage Room Second Floor" (real id) → `?location=is:<uuid>` → 1 row. Clicking "No Location" (sentinel) → `?location=is:without-location` → 1 row (the keyboard with null location). Both filter chips visible.

@coderabbitai review

File: apps/webapp/app/utils/cookies.server.ts [hunk 2]
```diff
@@ -221,6 +221,39 @@ export const createAdvancedAssetFilterCookie = (orgId: string) =>
...
+    if (!columnNames.includes(key)) {
+      normalized.append(key, value);
+      return;
+    }
+    if (advancedFilterFormatSchema.safeParse(value).success) {
+      normalized.append(key, value);
+      return;
+    }
+    if (value) {
+      normalized.append(key, `is:${value}`);
```
[the remark this answers is not in the document]
Comment (carlosvirreira) [reply apps/webapp/app/utils/cookies.server.ts]: Good catch — real regression. Reproduced your scenario locally: `?status=foo:AVAILABLE` with the round-1 normalizer in place was getting coerced to `is:foo:AVAILABLE` and then splitting downstream into operator=`is`, value=`foo`, which crashed the status enum cast.

Fixed in d8244c956. The normalizer now only promotes **unambiguous bare values** — a value containing `:` that fails operator validation is treated as a malformed operator attempt and dropped, matching the pre-PR drop-on-fail semantics. Added a regression test covering `?status=foo:AVAILABLE` (now 2005 vitest passing).

Verified end-to-end:
- Visiting `/assets?status=foo:AVAILABLE` in Advanced mode now redirects cleanly to the cookie's saved-filter state (no crash).
- Happy path: clicking a real-location slice still emits `?location=is:<uuid>` and filters correctly.

**Comment by carlosvirreira** [comment]
Round-3 pushed in d8244c956. One more Codex catch addressed:

**@chatgpt-codex-connector (P2) — "Drop malformed operator values instead of normalizing them"**

✅ Fixed. The round-1 normalizer was coercing values like `?status=foo:AVAILABLE` (malformed operator form, unknown operator `foo`) into `is:foo:AVAILABLE`. Downstream `parseFilters` then split it into operator=`is`, value=`foo`, and the status query crashed casting `foo` to the `AssetStatus` enum.

The normalizer now only promotes **unambiguous bare values**: a value that contains `:` and didn't validate as a known operator form is treated as a malformed operator attempt and dropped — preserving the pre-PR drop-on-fail semantics. Added a regression test (`drops malformed operator-form values instead of coercing them`).

**Status:** 2005 vitest passing, typecheck clean, both Codex review threads resolved. Manual verification:
- `/assets?status=foo:AVAILABLE` → redirects cleanly to cookie state, no crash.
- `/reports/distribution` → click real-location slice → `?location=is:<uuid>` → 1 row filtered (happy path intact).
- `/reports/distribution` → click "No Location" sentinel → `?location=is:without-location` → 1 row filtered (backend `IS NULL` translation, no defensive gate needed).

@coderabbitai review

**Comment by carlosvirreira** [comment]
@donkoko ready when you are — please merge.

Quick why this happened: a customer (UK IT supervisor, 5k assets / 207 locations) emailed me saying the new Asset Distribution donut links didn't filter the Assets page. Repro'd it: Advanced-mode workspaces hit a validator in `cookies.server.ts` that silently dropped any column filter without operator form (`is:<value>`) and then redirected to the cookie state — so the click looked like nothing happened. Simple-mode users never noticed. The fix is a normalize-don't-drop in that validator, plus a missed `locationId` parameter in `custodySnapshotReport` that the route was plumbing in but the helper was destructuring and ignoring.

Three review rounds processed (CodeRabbit nits + a Codex P2 catch on a malformed-operator edge case I introduced in round 1 and fixed in round 3). All threads resolved, CI all green, full `pnpm webapp:validate` passes, manual browser verification across 11 scenarios at HEAD covering both modes, sentinels, mixed valid/malformed URLs, cookie persistence, and Custody Snapshot.

Follow-ups not in this PR: #2542 (capability-gap design pass on the other reports), #2543 (Bookings inheritance constraint when it grows an Advanced mode), #2544 (React Doctor sticky-comment CI bug).

**Review by DonKoko — APPROVED** [review]
